File: src/audio/worklet/chroma.ts
// Chromagram (AGENTS.md §4.2/§4.5 step 4) — a 12-bin pitch-class energy
// vector, the one sense this pipeline had entirely missing before (no
// harmonic signal existed anywhere in Layer 1). Cheap, standard MIR: fold
// every FFT bin's energy into its nearest equal-tempered pitch class
// (0=C, 1=C#, ... 9=A, ... 11=B — MIDI-note-mod-12 convention, so A440 at
// MIDI 69 lands on pitch class 9), then normalize. Enables harmonic novelty
// (key/chord changes as a boundary cue — one of the EDM switch-point
// predictors §4.1 cites) and a genuinely new visual driver (pitch-class ->
// hue/rotation).
export const CHROMA_BINS = 12

// Bins outside a plausible musical range are skipped: sub-audio energy
// (hum, DC leakage) below C1 has no real pitch-class meaning and would
// dominate the accumulation (1/f-shaped spectra are bass-heavy), and
// anything above C8 is past where chroma is a meaningful timbral cue for
// this project's material.
const MIN_HZ = 32.7 // C1
const MAX_HZ = 4186 // C8

function freqToPitchClass(freqHz: number): number {
  // MIDI note number relative to A4=440Hz (MIDI 69); mod 12 with A4's own
  // class (9) already baked in by using 69 as the reference, so pitch class
  // 0 lands on C the way every other MIDI-based convention expects.
  const midi = 69 + 12 * Math.log2(freqHz / 440)
  const pc = Math.round(midi) % 12
  return pc < 0 ? pc + 12 : pc
}

// Reusable output buffer, same convention as the rest of this worklet
// (feature-worklet.ts writes into pre-allocated Float32Arrays every hop to
// avoid per-hop garbage) — pass one in rather than allocating here.
export function computeChroma(mags: Float32Array, sampleRate: number, fftSize: number, out: Float32Array): void {
  out.fill(0)
  const binHz = sampleRate / fftSize
  let total = 0
  for (let i = 1; i < mags.length; i++) {
    const freq = i * binHz
    if (freq < MIN_HZ || freq > MAX_HZ) continue
    const pc = freqToPitchClass(freq)
    out[pc] += mags[i]
    total += mags[i]
  }
  if (total > 1e-9) {
    for (let i = 0; i < CHROMA_BINS; i++) out[i] /= total
  }
}

// The pitch class with the most energy, as a 0..1 hue-ready value (not a
// hue itself — the caller picks the actual color mapping) — AGENTS.md
// §4.4's `chromaRootHue`. Returns 0 (arbitrary) for a silent/flat frame.
export function dominantPitchClassHue(chroma: Float32Array): number {
  let best = 0
  let bestValue = chroma[0] ?? 0
  for (let i = 1; i < chroma.length; i++) {
    if (chroma[i] > bestValue) {
      best = i
      bestValue = chroma[i]
    }
  }
  return best / CHROMA_BINS
}

File: crates/hyst-audio/src/dsp/chroma.rs
//! Chromagram, ported from `src/audio/worklet/chroma.ts`.

pub const CHROMA_BINS: usize = 12;

const MIN_HZ: f32 = 32.7; // C1
const MAX_HZ: f32 = 4186.0; // C8

fn freq_to_pitch_class(freq_hz: f32) -> usize {
    let midi = 69.0 + 12.0 * (freq_hz / 440.0).log2();
    let pc = midi.round().rem_euclid(12.0);
    pc as usize
}

/// Folds every FFT bin's energy into its nearest equal-tempered pitch class
/// (0=C .. 11=B), normalized. `out` must have length `CHROMA_BINS`.
pub fn compute_chroma(mags: &[f32], sample_rate: f32, fft_size: usize, out: &mut [f32]) {
    out.fill(0.0);
    let bin_hz = sample_rate / fft_size as f32;
    let mut total = 0.0f32;
    for (i, &m) in mags.iter().enumerate().skip(1) {
        let freq = i as f32 * bin_hz;
        if !(MIN_HZ..=MAX_HZ).contains(&freq) {
            continue;
        }
        let pc = freq_to_pitch_class(freq);
        out[pc] += m;
        total += m;
    }
    if total > 1e-9 {
        for v in out.iter_mut() {
            *v /= total;
        }
    }
}

/// The pitch class with the most energy, as a 0..1 hue-ready value.
pub fn dominant_pitch_class_hue(chroma: &[f32]) -> f32 {
    let mut best = 0;
    let mut best_value = chroma.first().copied().unwrap_or(0.0);
    for (i, &v) in chroma.iter().enumerate().skip(1) {
        if v > best_value {
            best = i;
            best_value = v;
        }
    }
    best as f32 / CHROMA_BINS as f32
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a440_lands_on_pitch_class_9_a() {
        // one bin exactly at 440Hz
        let fft_size = 2048;
        let sample_rate = 48_000.0;
        let bin_hz = sample_rate / fft_size as f32;
        let bin = (440.0 / bin_hz).round() as usize;
        let mut mags = vec![0.0; fft_size / 2 + 1];
        mags[bin] = 1.0;
        let mut chroma = vec![0.0; CHROMA_BINS];
        compute_chroma(&mags, sample_rate, fft_size, &mut chroma);
        let dominant = chroma
            .iter()
            .enumerate()
            .max_by(|a, b| a.1.partial_cmp(b.1).unwrap())
            .map(|(i, _)| i)
            .unwrap();
        assert_eq!(dominant, 9);
    }

    #[test]
    fn chroma_normalizes_to_sum_1() {
        let fft_size = 2048;
        let sample_rate = 48_000.0;
        let mut mags = vec![0.01; fft_size / 2 + 1];
        mags[300] = 5.0;
        let mut chroma = vec![0.0; CHROMA_BINS];
        compute_chroma(&mags, sample_rate, fft_size, &mut chroma);
        let sum: f32 = chroma.iter().sum();
        assert!((sum - 1.0).abs() < 1e-3);
    }

    #[test]
    fn dominant_hue_matches_the_loudest_bin() {
        let mut chroma = vec![0.0; CHROMA_BINS];
        chroma[4] = 0.9;
        assert_eq!(dominant_pitch_class_hue(&chroma), 4.0 / 12.0);
    }
}
